src/text/non_blocking/raw_text_reader.rs

```diff
@@ -25,6 +25,15 @@ use crate::IonType;
 
 const INITIAL_PARENTS_CAPACITY: usize = 16;
 
+// Represents reader state that can be carried over between calls to `next`.
+// Currently this is used to communicate state when `next` needs to return due to lack of data.
+enum ReaderState {
+    // A previous `next` call was able to parse the struct field's name, but not the value.
+    StructFieldValue,
```

Comment:
My initial thought here was that we should just make parsing a field name and value transactional; if either fails we just bubble up `Incomplete` and retry both on the next attempt.

We still need to track state for stepping out of a container, so we could keep this implementation even though it's a bit more complex. Interested to hear your thoughts, though.

Reply:
> What is the benefit of throwing the field name away? Is it just to keep the code simple?

Yep!

> Even if it is only during read boundaries, the data, and read/server response/etc size play into how impactful that is.

Intuitively, I'd expect the work wasted at read boundaries to be negligible in practice. I'd prefer to start simple and add additional state machinery if/when it shows up in profiling.